layers/cmd/pipeline.go

```diff
@@ -0,0 +1,85 @@
+package cmd
+
+import (
+	"github.com/cloudprivacylabs/lsa/layers/cmd/cmdutil"
+	"github.com/cloudprivacylabs/lsa/pkg/ls"
+	"github.com/cloudprivacylabs/opencypher/graph"
+	"github.com/spf13/cobra"
+)
+
+type PipelineContext struct {
+	Context     *ls.Context
+	Graph       graph.Graph
+	Roots       []graph.Node
+	InputFiles  []string
+	currentStep int
+	steps       []Step
+}
+
+type Step interface {
+	Run(*PipelineContext) error
+	Next() error
+}
+
+func (ctx *PipelineContext) Next() error {
+	ctx.currentStep++
+	if ctx.currentStep >= len(ctx.steps) {
+		return nil
+	}
+	err := ctx.steps[ctx.currentStep].Run(ctx)
+	ctx.currentStep--
+	return err
+}
+
+func init() {
+	ingestCmd.AddCommand(pipelineCmd)
+	pipelineCmd.Flags().String("file", "", "Pipeline build file")
+	pipelineCmd.Flags().String("initialGraph", "", "Load this graph and ingest data onto it")
+	// pipelineCmd.Flags().String("inputFile", "", "User provided input file")
+}
+
+var pipelineCmd = &cobra.Command{
+	Use:   "pipeline",
+	Short: "build pipeline JSON",
+	Args:  cobra.MaximumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		file, err := cmd.Flags().GetString("file")
+		if err != nil {
+			failErr(err)
+		}
+		// var steps []Step
+		var stages []map[string][]Step
+		err = cmdutil.ReadJSONOrYAML(file, &stages)
```

Comment:
This is not gonna work. You need to unmarshal to an intermediate struct

type stepMarshal struct {
   Operation string `json:"operation" yaml:"operation"`
   Step  json.RawMessage `json:"params" yaml:"params"`
}
1) Unmarshal into []stepMarshal
2) For each element in unmarshaled data, create Step based on Operation
3) Unmarshal into that Step from stepMarshal.Step

Put []Step into pipeline ctx

To map operations to structs:
```
var operations:=map[string]func() Step {
   "csvingest": func() Step {return &CSVIngest{}),
  ...
}
step:=operations[op]()
json.Unmarshal(params,step)

```


In every pipeline step .go file:

```
func init() {
   operations["csvingest"]=func() Step {return &CSVIngest}
}

```